narwhals/_pandas_like/expr.py

```diff
@@ -330,3 +342,117 @@ def total_microseconds(self) -> PandasExpr:
 
     def total_nanoseconds(self) -> PandasExpr:
         return register_namespace_expression_call(self._expr, "dt", "total_nanoseconds")
+
+
+class PandasExprNameNamespace:
+    def __init__(self, expr: PandasExpr) -> None:
+        self._expr = expr
+
+    def map(self, function: Callable[[str], str]) -> PandasExpr:
+        root_names = self._expr._root_names
+
+        if root_names is None:
+            msg = (
+                "Anonymous expressions are not supported in `.name.map`.\n"
+                "Instead of `nw.all()`, try using a named expression, such as "
+                "`nw.col('a', 'b')`\n"
+            )
+            raise ValueError(msg)
+
+        return self._expr.__class__(
+            lambda df: [
+                series.alias(function(series.name)) for series in self._expr._call(df)
+            ],
+            depth=self._expr._depth,
+            function_name=self._expr._function_name,
+            root_names=root_names,
+            output_names=[function(name) for name in root_names],
+            implementation=self._expr._implementation,
+        )
```

Comment:
I will comment just one but logic applies for all the methods.

- This implementation works fine with chaining operations (cfr updated test suite)
- I am now calling `self._expr._call` and renaming the series using alias method, and name attribute
- The renaming is done on the `root_names`
- Unsure if the depth should be increased or not

Reply:
nice!

depth shouldn't be increased, as this is still just a simple aggregation, so we can use the fastpath in group_by

I'm not really proud of using `depth` to determine whether an aggregation is simple in group_by...maybe we can make a `SimplePandasExpr` subclass of `PandasExpr`, and then do an `isinstance` check...that should be clearer to read and to maintain

Reply:
Not sure I completely follow along. I will need to come back at this 😂

Reply:
I'll  explain it in the 'how it works' PR - anyway, I think you're doing things correctly here 👍 

one way to check would be to do something like
```python
import pandas as pd
import narwhals as nw
dfpd = pd.DataFrame({'a': [1,1,2], 'b': [5,6,7]})
df = nw.from_native(df)
df = df.group_by('a').agg(pl.col('b').name.suffix('_baz').mean())
print(nw.to_native(df))
```
and check that:
1. you get the correct result
2. no perf warning is raised

Reply:
If I add an _aggregation_ I get a warning which seems unrelated.

```py
import pandas as pd
import narwhals as nw
df_pd = pd.DataFrame({'a': [1,1,2], 'b': [5,6,7]})
df = nw.from_native(df_pd)
df = df.group_by('a').agg(nw.col('b').mean().name.suffix('_baz').mean())
print(nw.to_native(df))
```
results in:
```terminal
 .../narwhals/narwhals/_pandas_like/group_by.py:56: UserWarning: Found complex group-by expression, which can't be expressed efficiently with the pandas API. If you can, please rewrite your query such that group-by aggregations are simple (e.g. mean, std, min, max, ...).
  return agg_pandas(
   a  b_baz
0  1    5.5
1  2    7.0
```

Reply:
thanks for checking - looks fine then, the warning looks correct if there's a double-aggregation. will expand on it in the other PR 😇